btcec/schnorr/musig2/musig2_test.go

```diff
@@ -1443,6 +1444,228 @@ func (mr *memsetRandReader) Read(buf []byte) (n int, err error) {
+		{
+			partialSigs: [][]byte{
+				psig7,
+				mustParseHex("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141"),
+			},
+			aggNonce:      combineSigAggNonce3,
+			expectedError: ErrPartialSigInvalid,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var pSigs []*PartialSignature
+		for _, partialSig := range testCase.partialSigs {
+			pSig := &PartialSignature{}
+			err := pSig.Decode(bytes.NewReader(partialSig))
+			if err != nil {
+				if errors.Is(err, testCase.expectedError) {
+					return
+				}
+				t.Fatal(err)
+			}
+			pSigs = append(pSigs, pSig)
+		}
+
+		keySet := make([]*btcec.PublicKey, 0, len(testCase.keyOrder))
+		for _, keyIndex := range testCase.keyOrder {
+			keyBytes := combineSigKeys[keyIndex]
+			pub, err := schnorr.ParsePubKey(keyBytes)
+			if err != nil {
+				t.Fatalf("unable to parse pubkeys: %v", err)
+			}
+			keySet = append(keySet, pub)
+		}
+
+		uniqueKeyIndex := secondUniqueKeyIndex(keySet, false)
+		aggOpts := []KeyAggOption{
+			WithUniqueKeyIndex(uniqueKeyIndex),
+		}
+		if len(testCase.tweaks) > 0 {
+			aggOpts = append(aggOpts, WithKeyTweaks(testCase.tweaks...))
+		}
+
+		combinedKey, _, _, err := AggregateKeys(
+			keySet, false, aggOpts...,
+		)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		aggPubkey, err := aggNonceToPubkey(testCase.aggNonce, combinedKey, combineSigsMsg)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		var opts []CombineOption
+		if len(testCase.tweaks) > 0 {
+			opts = append(opts, WithTweakedCombine(combineSigsMsg, keySet, testCase.tweaks, false))
+		}
+
+		sig := CombineSigs(aggPubkey, pSigs, opts...)
+		expectedSig, err := schnorr.ParseSignature(testCase.expected)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !bytes.Equal(sig.Serialize(), expectedSig.Serialize()) {
+			t.Fatalf("sigs not expected %x \n got %x", expectedSig.Serialize(), sig.Serialize())
+		}
+
+		if !sig.Verify(combineSigsMsg[:], combinedKey.FinalKey) {
+			t.Fatal("sig not valid for m")
+		}
+	}
+}
+
+// aggNonceToPubkey gets a nonce as a public key for the TestMusig2CombineSigsTestVectors
+// test.
+func aggNonceToPubkey(combinedNonce [66]byte, combinedKey *AggregateKey, msg [32]byte) (*btcec.PublicKey, error) {
+
+	// b = int_from_bytes(tagged_hash('MuSig/noncecoef', aggnonce + bytes_from_point(Q) + msg)) % n
+	var (
+		nonceMsgBuf  bytes.Buffer
+		nonceBlinder btcec.ModNScalar
+	)
+	nonceMsgBuf.Write(combinedNonce[:])
+	nonceMsgBuf.Write(schnorr.SerializePubKey(combinedKey.FinalKey))
+	nonceMsgBuf.Write(msg[:])
+	nonceBlindHash := chainhash.TaggedHash(NonceBlindTag, nonceMsgBuf.Bytes())
+	nonceBlinder.SetByteSlice(nonceBlindHash[:])
+
+	r1, err := btcec.ParsePubKey(
+		combinedNonce[:btcec.PubKeyBytesLenCompressed],
+	)
+	if err != nil {
+		return nil, err
+	}
+	r2, err := btcec.ParsePubKey(
+		combinedNonce[btcec.PubKeyBytesLenCompressed:],
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var nonce, r1J, r2J btcec.JacobianPoint
```

Comment:
Should we use this as an intermediate routine for the normal signing/verification stuff?

Reply:
I'm not sure I follow. Do you mean change the signature of `CombineSigs`?

Reply:
I mean reducing duplication by moving some of the above into a new routine, so we can use it elsewhere. Don't consider it a blocker though, IIRC there's an existing TODO lingering around somewhere for it